src/sentry/models/outbox.py

```diff
@@ -524,28 +503,65 @@ def save(self, **kwds: Any) -> None:  # type: ignore[override]
         metrics.incr("outbox.saved", 1, tags=tags)
         super().save(**kwds)
 
+    def lock_id(self, attrs: Iterable[str]) -> int:
+        # 64 bit integer that roughly encodes a unique, serializable lock identifier
+        return mmh3.hash64(".".join(str(getattr(self, attr)) for attr in attrs))[0]
+
     @contextlib.contextmanager
     def process_shard(
-        self, latest_shard_row: OutboxBase | None
+        self,
+        latest_shard_row: OutboxBase | None,
+        lock_timeout: int = 5,
     ) -> Generator[OutboxBase | None, None, None]:
-        flush_all: bool = not bool(latest_shard_row)
-        next_shard_row: OutboxBase | None
         using: str = db.router.db_for_write(type(self))
-        with transaction.atomic(using=using), django_test_transaction_water_mark(using=using):
+
+        shard_lock_id = self.lock_id(self.sharding_columns)
+        with connections[using].cursor() as cursor:
+            cursor.execute("show lock_timeout")
+            orig_timeout = cursor.fetchone()[0]
+            cursor.execute(f"SET local lock_timeout='{lock_timeout}s'")
```

Comment:
Have you considered using `sentry.locks` instead of postgres advisory locks? While `sentry.locks` uses redis under the hook it also features ways to automatically retry lock acquisition when there is contention. One big downside to `sentry.locks` is that because it is redis backed it cannot participate in transactions, but it doesn't seem like we're doing transactions here.

Reply:
I believe Zach mentioned that he considered this, but postgres locks were a better match because they die at the same time that the postgres connection dies, while redis locks could easily outlive the worker that created them... I think that was what he said :)

Reply:
Correct, you can talk to @mrduncan about further details as we discussed at length our options before arriving at postgresql locks.